### PR TITLE
Atomic.util.image_by_name only matches some possible matches

### PR DESCRIPTION
**Pull request by ibotty**
e.g. the following image (in docker)

    docker.io/user/image:tag

is not matched by the following queries, because the first `/` might not separate the registry from the user-image but the user from the image. 

 * user/image
 * image

The fix for that case is straightforward, but the catch is, that non-prefixed containers (e.g. docker.io/fedora:latest) show an ambiguity: It could be user `docker.io` or registry `docker.io`.

Non-user supplied queries to `_decompose` might also be incorrectly matched, because (at least on one of my machines) there were images without registry: `user/image:latest` in docker's images output.

The patch fixes it by not decomposing the query (and docker's images), but directly matching four times.

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM

@baude @cgwalters @aweiteka PTAL

**Comment by rhatdan** [comment]
@bexelbie @baude @cgwalters @aweiteka @miminar PTAL

**Comment by miminar** [comment]
I fear this would produce a lot of ambiguity. `image_by_name` is used inside functions that expect single image to match and will raise otherwise.

And to prevent user confustion, I'd vote to use same image matching mechanisms as Docker itself. Upstream's `docker images` wouldn't match `registry.ltd/user/image` with just `*/image` either.

**Comment by bexelbie** [comment]
I don't think a user/image should match without the query containing "user/image"

Could we improve the repo detection logic by checking for a '.' in the item preceding the slash? It appears that usernames are not allowed to have non-alphanumerics.

**Comment by rhatdan** [comment]
@ibotty Could you address the issues above?

**Comment by ibotty** [comment]
Oh, I missed the last comments. Sorry for not doing anything about it.

I am not entirely sure what should be matched by what. Maybe a single
function won't suffice.

I don't think we should mimick `docker images` behavior. In the list
below, it only matches the last query. That can't be what we want.

So, given an image registry/user/image:tag, which of the following queries
should succeed?

 * image
 * image:tag
 * user/image
 * user/image:tag
 * registry/user/image:tag
 * registry/user/image

if user is indeed [a-zA-Z0-9], and repos always contain a `.`, it should
be easy to do.

I might get to it later that week.

**Comment by jlebon** [comment]
bot, test pull request

**Comment by jlebon** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
@ibotty Any update on this?

**Comment by ibotty** [comment]
I am still not sure about the semantics we want. Can someone please comment on my comment above? projectatomic/atomic wrote:@ibotty Any update on this?

**Comment by rhatdan** [comment]
@miminar @bexelbie PTAL

**Comment by rhatdan** [comment]
@baude @runcom PTAL

**Comment by rhatdan** [comment]
@ibotty Have you thought more on this?  Or just dropped it.

**Comment by ibotty** [comment]
I am happy  to implement any semantics you want to have. Having my comment above in mind, I would say the image should be matched by image/user and everything below. What do you say?

**Comment by ibotty** [comment]
typo: user/image

**Comment by rhatdan** [comment]
Sounds good to me.

**Comment by rhatdan** [comment]
@ibotty Have you had time to work on this?

**Comment by ibotty** [comment]
Sorry, I did not. No ETA, because I can't do it in work time. It's no priority anymore unfortunately.

When is 1.10 due? Maybe I can dedicate some time today in the middle European evening to get it in in time.

**Comment by ibotty** [comment]
Please have a look at it now. The fix was made way easier by @bexelbie's comment above, that registries always contain a '.'.

**Comment by rhatdan** [comment]
LGTM

@baude @runcom PTAL

**Comment by rhatdan** [comment]
@cevitch Can you fix the jenkens or know who can?

**Comment by rhatdan** [comment]
@cgwalters Do you know who can fix the jenkins?

**Comment by cgwalters** [comment]
bot, retest this please

**Comment by jlebon** [comment]
@rhatdan 

There are two unit tests failing:

```
UNIT TESTS:
.........FF
======================================================================
FAIL: test_image_by_name_registry_match (test_util.TestAtomicUtil)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/atomic-upstream-test/atomic/tests/unit/test_util.py", line 26, in test_image_by_name_registry_match
    self.assertTrue(len(matches) == 1)
AssertionError: False is not true

======================================================================
FAIL: test_image_by_name_tag_glob (test_util.TestAtomicUtil)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/atomic-upstream-test/atomic/tests/unit/test_util.py", line 22, in test_image_by_name_tag_glob
    self.assertTrue(len(matches) == 1)
AssertionError: False is not true

----------------------------------------------------------------------
```

I haven't tried reproducing this locally yet, though it does seem strange since the tester does clear out all images before each test.

**Comment by ibotty** [comment]
I am pretty sure it's the leading `/` in the image to be matched. I did not test it with it. One could simply drop a potential leading slash in that case.

**Comment by ibotty** [comment]
I think I fixed the test failure. If someone can retest it please.

File: Atomic/util.py [hunk 1]
```diff
@@ -26,9 +26,12 @@
 
 def _decompose(compound_name):
     """ '[reg/]repo[:tag]' -> (reg, repo, tag) """
-    reg, repo, tag = '', compound_name, ''
+    reg, repo, tag = '', lstrip(compound_name, '/'), ''
```
Comment (jlebon) [inline Atomic/util.py]: Do you mean `compound_name.lstrip('/')` ?

Comment (ibotty) [inline Atomic/util.py]: I am spending too much time in functional programming languages. I certainly do.

**Comment by rhatdan** [comment]
@ibotty Rebase please and then we can merge if tests pass.

**Comment by rhatdan** [comment]
@baude could you check if this is something we need at this point, or have we already implemented it?

**Comment by baude** [comment]
@rhatdan i dont have any context on this anymore.

@ibotty any chance you can provide some examples I could use to test master and then your code?  examples could just be images names, etc

**Comment by ibotty** [comment]
@baude I lost context as well. But afair, just e.g. docker.io/fedora/apache:latest with queries fedora/apache should exhibit the problem/the fix.

**Comment by rhatdan** [comment]
I am closing this since I think it no longer applies.

**Pull request closed without merging**